### PR TITLE
[FLINK-9410] [yarn] Replace NMClient with NMClientAsync in YarnResour…

### PR DESCRIPTION
**Pull request by rice904**
## What is the purpose of the change
YarnResourceManager uses the synchronous ```NMClient``` which is called from within the main thread of the ```ResourceManager```. Since these operations are blocking, we should replace the client with the ```NMClientAsync``` and make the calls non blocking.

## Brief change log
Replace NMClient with AsyncNMClient. And change the test also.

## Verifying this change
Run YarnResourceManagerTest verify this.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no )
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by sihuazhou** [comment]
Hi @zhangminglei , do you want to take over this? if yes, plz free feel to take over this. I had a brief look at the code, I think this may not as easy as replace all api() to apiAsync(), e.g. maybe you also need to do some retry things in the `onStartContainerError()` of `NMClientAsync`.

**Comment by zhangminglei** [comment]
Hi, @sihuazhou . Thank you for your feedback. I would do not think retry here since error has occurred. we should tell the user the exception. I am not very sure the corresponding error whether need retry. @tillrohrmann  Could you take also look on this PR ? Thanks .

**Comment by sihuazhou** [comment]
Hi @zhangminglei , I would suggest to retry, this is more consistent with the behavior when we are using the sync API, in fact, currently if we don't retry to start container then the job will fail to acquire slots, and RM will also stop to start container for the jobs that are waiting for slots.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 1]
```diff
@@ -510,4 +509,39 @@ private int generatePriority(ResourceProfile resourceProfile) {
 		}
 	}
 
+	/**
+	 * Implements NMClientAsyncCallbackHandler.
+	 */
+	private class NMClientAsyncCallbackHandler implements NMClientAsync.CallbackHandler {
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Maybe we could let YarnResourceManager to implement this interface, as it has implemented the `AMRMClientAsync.CallbackHandler`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 2]
```diff
@@ -510,4 +509,39 @@ private int generatePriority(ResourceProfile resourceProfile) {
...
+		}
+
+		@Override
+		public void onContainerStatusReceived(ContainerId containerId, ContainerStatus containerStatus) {
+
+		}
+
+		@Override
+		public void onContainerStopped(ContainerId containerId) {
+
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: I would suggest to add some log here.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 3]
```diff
@@ -212,13 +214,12 @@ private void getContainersFromPreviousAttempts(final RegisterApplicationMasterRe
...
-		NMClient nodeManagerClient = NMClient.createNMClient();
-		nodeManagerClient.init(yarnConfiguration);
-		nodeManagerClient.start();
-		nodeManagerClient.cleanupRunningContainersOnStop(true);
-		return nodeManagerClient;
+	protected NMClientAsync createAndStartNodeManagerClient(YarnConfiguration yarnConfiguration) {
+		// create the client to communicate with the node managers.
+		NMClientAsync nodeManagerAsyncClient = NMClientAsync.createNMClientAsync(new NMClientAsyncCallbackHandler());
+		nodeManagerAsyncClient.init(yarnConfiguration);
+		nodeManagerAsyncClient.start();
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Maybe we should also call this `nodeManagerClient.getClient().cleanupRunningContainersOnStop(true);`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 1 above)
Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Yea. Will change.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 2 above)
Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Okay. Will change.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 3 above)
Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: I am not very sure about this. But I will change it now.

**Comment by tillrohrmann** [comment]
Thanks for opening the PR @zhangminglei. I agree with @sihuazhou that it's not as easy as replacing `x` with `xAsync`. As @sihuazhou pointed out, we have to react to asynchronous Yarn messages depending on our internal state. So for example, when calling `startContainerAsync` we should store somewhere that we tried to start a container. If this request later fails, we should not fail the complete ResourceManager but rather send a new container request (as it is done in the `YarnResourceManager#onContainersAllocated` method).

Additionally, while touching this code, we should check whether we can improve the unit tests for this component. Especially with the asynchronous node manager client, it should be rather simple to write good tests when creating a mock implementation and manually calling the callbacks.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 4]
```diff
@@ -356,7 +363,8 @@ public void onContainersAllocated(List<Container> containers) {
 
 					final String containerIdStr = container.getId().toString();
 					final ResourceID resourceId = new ResourceID(containerIdStr);
-
+					this.resourceId = resourceId;
+					this.container = container;
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: I think we can't do it like this, the `resourceId` will be reset by another request... as well as the `contianer`

Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Yes. Thanks, I know what is wrong with my code.Will change.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 5]
```diff
@@ -406,6 +407,43 @@ public void onError(Throwable error) {
...
+	public void onContainerStopped(ContainerId containerId) {
+		log.info("The container {} is stopped.", containerId);
+	}
+
+	@Override
+	public void onStartContainerError(ContainerId containerId, Throwable t) {
+		log.error("Could not start TaskManager in container {}.", containerId, t);
+
+		// release the failed container
+		YarnWorkerNode yarnWorkerNode = workerNodeMap.remove(new ResourceID(containerId.toString()));
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Maybe it safer to check whether yarnWorkerNode is null.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 6]
```diff
@@ -406,6 +407,43 @@ public void onError(Throwable error) {
...
+		resourceManagerClient.releaseAssignedContainer(containerId);
+		// ask for a new one
+		requestYarnContainer(yarnWorkerNode.getContainer().getResource(), yarnWorkerNode.getContainer().getPriority());
+	}
+
+	@Override
+	public void onGetContainerStatusError(ContainerId containerId, Throwable t) {
+		log.error("Error occurred during get the container {} status.", containerId, t);
+		onFatalError(t);
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: I think maybe this is not a fatal error.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 7]
```diff
@@ -406,6 +407,43 @@ public void onError(Throwable error) {
...
+	}
+
+	@Override
+	public void onStopContainerError(ContainerId containerId, Throwable t) {
+		log.error("Error occurred during stop the container {}.", containerId, t);
+		onFatalError(t);
```
Comment (sihuazhou) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: some above.

Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: change.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 6 above)
Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: change

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 5 above)
Comment (zhangminglei) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Will change.

**Comment by zhangminglei** [comment]
Hi, @tillrohrmann  Could you also take another look on this PR ? If the code wrong, please let me know. and I will give a quick fix. Thank you!

**Comment by tillrohrmann** [comment]
Closing because the code base has diverged too much from this PR.

**Pull request closed without merging**